src/openforms/registrations/api/views.py

```diff
@@ -86,13 +123,48 @@ def get_objects(self):
             catalogus["url"]: catalogus for catalogus in catalogus_data
         }
 
-        iotypen_data = client.get_all_informatieobjecttypen()
-        iotypen = [
-            {
-                "informatieobjecttype": factory(InformatieObjectType, iotype),
-                "catalogus": factory(Catalogus, catalogus_mapping[iotype["catalogus"]]),
-            }
-            for iotype in iotypen_data
-        ]
+        if "catalogus_domein" in filter_serializer.validated_data:
+            domein = filter_serializer.validated_data["catalogus_domein"]
+            # `catalogus_rsin` is guaranteed to be present if `catalogus_domein` is:
+            rsin = filter_serializer.validated_data["catalogus_rsin"]
+
+            matching_catalogus: str | None = next(
+                (
+                    catalogus["url"]
+                    for catalogus in catalogus_data
+                    if catalogus["domein"] == domein and catalogus["rsin"] == rsin
+                ),
+                None,
+            )
+
+            if matching_catalogus is None:
+                return []
+
+            iotypen_data = client.get_all_informatieobjecttypen(
+                catalogus=matching_catalogus
+            )
+        else:
+            iotypen_data = client.get_all_informatieobjecttypen()
```

Comment:
can you add a comment to describe in which situations/code-paths we *don't* filter based on catalogus yet? That way we have a list of things to still improve to clean up this branching.

Reply:
I think https://github.com/open-formulieren/open-forms/pull/4507#discussion_r1677592130 simplifies things already